SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialSpanIdStack.cpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/Connection/SpatialSpanIdStack.h"
+
+namespace SpatialGDK
+{
+void FSpatialSpanIdStack::AddNewLayer(const Trace_SpanId& SpanId)
+{
+	Stack.Add({ SpanId });
+}
+
+void FSpatialSpanIdStack::AddToLayer(const Trace_SpanId& SpanId)
+{
+	const int32 Size = Stack.Num();
+	if (Size == 0)
+	{
+		AddNewLayer(SpanId);
+		return;
+	}
+
+	Stack[Size - 1].Add(SpanId);
+}
+
+TArray<Trace_SpanId> FSpatialSpanIdStack::PopLayer()
+{
+	const int32 Size = Stack.Num();
+	const TArray<Trace_SpanId> TopLayer = Stack[Size - 1];
```

Comment:
```suggestion
	const TArray<Trace_SpanId> TopLayer = MoveTemp(Stack[Size - 1]);
```

Reply:
I don't think we should add MoveTemp here.
The reason we use MoveTemp is to avoid copy construction, but here the return type is TArray<Trace_SpanId>, even we use MoveTemp the TopLayer has construction too.
The first time, I prefer using
`TArray<Trace_SpanId>& TopLayer = Stack[Size - 1];`
It means the `TopLayer` is a reference and can save 1-time construction and destruction.

But In my opinion, return TArray<Trace_SpanId> is not a good choice.
Maybe change it to 
```
bool FSpatialSpanIdStack::PopLayer(TArray<Trace_SpanId>& SpanIds)
{

}
```
is better?

Reply:
I don't think we can use a reference as we delete the TArray<Trace_SpanId> element in this line: Stack.RemoveAt(Size - 1). Perhaps I am wrong here?  With the MoveTemp I think we remove all constructions. I believe part of the c++ standard (perhaps 11) means that there will not be extra copies when returning a value but I can't remember what it is called 🤔 Perhaps it's called "return value optimization".

Reply:
Perhaps I will just call .Pop() on the array

Reply:
Here you have a temporary variable TopLayer to have it and return it. I think it's ok.
If we are very concerned about the performance of this code, reducing the number of constructor calls is necessary, but if not, the original way is fine.

Reply:
I have change the code to use .Pop() on the array which is simpler. Indeed there would be an extra constructor call for the temporary variable if we used on old c++ standard. I believe a newer feature in the new c++ standard which can be relied upon removes the unnecessary constructor call.

Reply:
nice job! use Pop is a more simple way to do it.